Fix route quotas calculation

alain-andre opened the pull request.

alain-andre committed:
fix route quotas calculation - close #137

alain-andre committed:
Fix route location calculation

senhalil commented on api/v01/api_base.rb:
```diff
@@ -41,11 +41,12 @@ def self.count_locations(obj)
 
-      def self.count_route_locations(params)
+      # Calculate route for given points considering start and end are not the same
+      def self.count_routes(params)
```
`count_route_locations` was better but `count_route_legs` would be even better IMHO because we are not counting the "routes", we are counting the "legs" of the routes. https://en.termwiki.com/EN/leg_(route) So route with A, B, C, D locations would have 3 legs (A-B, B-C, C-D)..

alain-andre replied:
Indeed `count_route_legs` is better.

senhalil commented on api/v01/api_base.rb:
```diff
@@ -41,11 +41,12 @@ def self.count_locations(obj)
       end
 
-      def self.count_route_locations(params)
+      # Calculate route for given points considering start and end are not the same
```
There is no calculation here, did you mean "*Count* route *legs* for given ..."

alain-andre replied:
Why do you says there is no calculation, it does a `count_locations(params[:loc]) - 1`

senhalil replied:
My comment was about the code comment (#) itself: we don't "calculate routes" but we "count the route legs"

senhalil commented on api/v01/api_base.rb:
```diff
@@ -41,11 +41,12 @@ def self.count_locations(obj)
+          count_locations(params[:loc]) - 1
         elsif params[:locs]
-          count_locations(params[:locs])
+          count_locations(params[:locs]) - 1
```
I am not very familiar with `routes` endpoint. Do we treat multiple routes in the same request? If so, this should be `-number_of_separate_routes` instead of `-1` because there is no leg between different routes, right? Does it make sense, like I said I am not very familiar with this endpoint, if my remark is incomprehensible it might be because I am way off :rocket:

alain-andre replied:
There is different way to count it. 

  - n-1 for a type A-B-C-D
  - n*(n-1)/2 for a type A-B-C-A
  
And an other when it's like A-B D-E-F

senhalil replied:
I didn't get why `n*(n-1)/2` for `A-B-C-A` it is still `n-1` with `n=4`.

For `A-B    D-E-F` it equals to 3 -- i.e., `2-1 + 3-1` which equals to `n - m` where `m` is the number of routes with `n=5`(2+3) and `m=2`

senhalil replied:
Note that `n-m` is the general form and it works perfectly for the single route case as well since `m=1` => `n-1`

alain-andre replied:
Yes, thanks for the formula. `n*(n-1)/2 ` works for triangles, not our use case. 

I have modified the code using this formula.

senhalil commented on api/v01/api_base.rb:
```diff
@@ -41,11 +41,12 @@ def self.count_locations(obj)
+      # Calculate route for given points considering start and end are not the same
+      def self.count_routes(params)
         if params[:loc]
-          count_locations(params[:loc])
+          count_locations(params[:loc]) - 1
         elsif params[:locs]
```
It looks like `params[:loc]` is always transformed into `params[:locs]` if this is systematic, do we need this if/elsif logic?

senhalil replied:
Also I am not so sure about this but if there is no `:loc` or `:locs` wouldn't it be better to return `0` instead of `nil`? Because in an empty route there are `0` legs this would prevent `cannot do comparison with nil` type of errors.

alain-andre replied:
You can send a query with `:loc` or `:locs`, it's an inherited thing. For the comparison with nil, `count_locations` returns 0 if no loc is given, I have fixed it.

senhalil replied:
I resolves this since fixing `:loc`/`:locs` duality is out of this PR's scope but it you think it's worth you can create a ticket to uniformise this "duality" so that we have one single implementation.

senhalil replied:
Again out of scope of this PR: I also realized that almost all of the endpoints start with a boiler-plate almost copy-paste like very similar logic that we can uniformise and extract into a single function so that we maintain less code and the  modification are uniform across all endpoints.

senhalil commented on api/v01/api_base.rb:
```diff
@@ -60,6 +66,41 @@ def self.limit_matrix_side_size(params)
+        when String
+          obj.split('\;|\|').map { |locs| count_locations(locs) }.sum
+        end
+      end
```
I don't understand why we need this function exactly. If I understand it correctly `count_locations` already handles this case correctly. We just need to count the number of distinct routes..

senhalil commented on api/v01/api_base.rb:
```diff
@@ -36,16 +36,22 @@ def self.count_locations(obj)
+      # For A-B D-E-F it equals to 3 -- i.e., 2-1 + 3-1 => 2+3 - 1+1
+      def self.count_route_legs(params)
         if params[:loc]
-          count_locations(params[:loc])
+          count_locations(params[:loc]) - 1
         elsif params[:locs]
-          count_locations(params[:locs])
+          [1, legs_in_distinct_routes(params[:locs]) - distinct_routes(params[:locs])].max
         end
       end
```
All we need is a function that does the following:

```ruby
def def self.count_route_legs(params)
  location = params[:loc] || params[:locs]
  count_locations(location) - distinct_route_count(locations)
end
```

where `distinct_route_count` counts the number of routes simply as follows
```ruby
def distinct_route_count(locations)
  case locations
  when String
    locations.count('\;|\|') + 1 # (i.)
  when Array
    locations.size  # (ii.)
  end
end
```
@alain-andre  I need your help to simplify `distinct_route_count`:

- (i.) I am not sure if `|` or `;` mean a new "route" or just a new location in the list, it is not possible understand from the given examples. If `|` or `;` means new route then the given examples do not make much sense because `"lat,lng;lat,lng" ` and  `"lat,lng|lat,lng"` both lead to two separate routes with 0 legs (1 locations). 

```
          # route, matrix and isoline can send value like :
          # "lat,lng,lat,lng" or "lat,lng;lat,lng" or "lat,lng|lat,lng"
```

- (ii.) Also if `depth=3` is multiple routes, then what is `depth=2`? For example what is the meaning of `[[lat, lng,lat, lng,lat, lng], [lat, lng,lat, lng]]`? Is it legitimate? Since `depth=1` is definetely single route, I would expect to that `depth=2` is multi-route. Above I assumed (as you noted as a comment) that only `depth==3` can be multi-route but if not we should fix it.

```
         # route, matrix and isoline can send array like :
          # [[[lat, lng], [lat, lng]]] or [[lat, lng], [lat, lng]] or [lat, lng, lat, lng]
```

@braktar do you know the exact syntax for (i.) and (ii.)?

braktar replied:
(i.) Both `|` and `;` are route separators (note `|` doesn't work with postman).  
I confirm that nor `"lat,lng|lat,lng"` nor `"lat,lng;lat,lng"` works it returns `"locs: segment #1, at least two couples of lat/lng required.`

(ii.) I have no opinion on this point

senhalil replied:
Cheers! We will look into this with Alain tomorrow or Friday and your observations already helped with half of the investigation.

I guess we can do as you did and use insomnia to figure out if and what  `[[lat, lng,lat, lng,lat, lng], [lat, lng,lat, lng]]` means?

senhalil replied:
it turns out it is not even possible to send `Array` `[]` and the chang is done internally and the API endpoint is based on string. So I simplified my suggestion accordingl

senhalil reviewed (changes requested):
It's getting there but there is an a lot simpler way to `count_route_legs` :point_down:

senhalil commented on api/v01/api_base.rb:
```diff
@@ -36,16 +36,22 @@ def self.count_locations(obj)
         else
-          obj.split(',').size / 2 # matrix, isoline, route : "lat,lng,lat,lng"
+          # route, matrix and isoline can send value like :
+          # "lat,lng,lat,lng" or "lat,lng;lat,lng" or "lat,lng|lat,lng"
```
```suggestion
          # "lat,lng,lat,lng,lat,lng" -> one route with 3 locations
          # "lat,lng,lat,lng,lat,lng;lat,lng,lat,lng" -> two routes with 3+2 locations
          # "lat,lng,lat,lng,lat,lng|lat,lng,lat,lng" -> two routes with 3+2 locations
```

senhalil commented on api/v01/api_base.rb:
```diff
@@ -36,16 +36,22 @@ def self.count_locations(obj)
           # route, matrix and isoline can send array like :
           # [[[lat, lng], [lat, lng]]] or [[lat, lng], [lat, lng]] or [lat, lng, lat, lng]
```
```suggestion
          # For route(s), loc(s) becomes an array of arrays internally as follows:
          # [[[lat, lng], [lat, lng], [lat, lng]]] -> 1 route with 3 locations 
          # [[[lat, lng], [lat, lng], [lat, lng]], [[lat, lng], [lat, lng], [lat, lng]]] -> 2 routes with 3 locations each
          # For matrix, src and dst become internally array and array of arrays:
          # [lat, lng, lat, lng, lat, lng] (this is the case for isoline too) 
          # [[lat, lng], [lat, lng], [lat, lng]]
```

senhalil committed:
Let functions manage their own extreme cases

Improve comments and error handling

count_locations
count_route_legs
count_matrix_cells

senhalil committed:
Simplify count_route_legs logic

senhalil committed:
Fix broken tests

alain-andre commented:
@braktar can you review this PR ?
